src/main/java/org/squiddev/plethora/integration/astralsorcery/MetaItems.java

```diff
@@ -0,0 +1,365 @@
+
+			IConstellation constellation = item.getFocusConstellation(stack);
+			if (constellation != null) {
+				out.put("constellationFocus", context.makePartialChild(constellation).getMeta());
+			}
+
+			IConstellation trait = ItemTunedCrystalBase.getTrait(stack);
+			if (trait != null) {
+				out.put("constellationTrait", context.makePartialChild(trait).getMeta());
+			}
+
+			return out;
+		}
+
+		@Nonnull
+		@Override
+		public ItemStack getExample() {
+			ItemStack crystalStack = new ItemStack(ItemsAS.tunedRockCrystal);
+			CrystalProperties.applyCrystalProperties(crystalStack, CrystalProperties.getMaxRockProperties());
+			ItemTunedCrystalBase.applyMainConstellation(crystalStack, Constellations.discidia);
+			ItemTunedCrystalBase.applyTrait(crystalStack, Constellations.gelu);
+
+			return crystalStack;
+		}
+	};
+
+	public static final IMetaProvider<ItemStack> META_CONSTELLATION_PAPER = new ItemStackContextMetaProvider<ItemConstellationPaper>(
+		ItemConstellationPaper.class,
+		"Provides the Constellation for a Constellation Paper"
+	) {
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<ItemStack> context, @Nonnull ItemConstellationPaper item) {
+			IConstellation constellation = ItemConstellationPaper.getConstellation(context.getTarget());
+			return constellation != null
+				? Collections.singletonMap("constellation", context.makePartialChild(constellation).getMeta())
+				: Collections.emptyMap();
+		}
+
+		@Nonnull
+		@Override
+		public ItemStack getExample() {
+			ItemStack paperStack = new ItemStack(ItemsAS.constellationPaper);
+
+			ItemConstellationPaper.setConstellation(paperStack, Constellations.discidia);
+
+			return paperStack;
+		}
+	};
+
+	public static final IMetaProvider<ItemStack> META_ITEM_JOURNAL = new ItemStackContextMetaProvider<ItemJournal>(
+		ItemJournal.class,
+		"Provides the Constellations stored in a Journal"
+	) {
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<ItemStack> context, @Nonnull ItemJournal item) {
+			//Just getting the constellations, rather than the papers; removes needless indirection,
+			// and reduces processing overhead (Astral's `getStoredConstellationStacks` calls `getStoredConstellations`,
+			// and then creates an ItemStack for each)
+			return Collections.singletonMap("papers",
+				LuaList.of(ItemJournal.getStoredConstellations(context.getTarget()),
+					paper -> context.makePartialChild(paper).getMeta()).asMap());
+		}
+
+		@Nonnull
+		@Override
+		public ItemStack getExample() {
+			ItemStack stack = new ItemStack(ItemsAS.journal);
+
+			LinkedList<IConstellation> constellations = new LinkedList<>();
+			constellations.add(Constellations.discidia);
+
+			ItemJournal.setStoredConstellations(stack, constellations);
+
+			return stack;
+		}
+	};
+
+
+	public static final IMetaProvider<ItemStack> META_SKY_RESONATOR = new ItemStackContextMetaProvider<ItemSkyResonator>(
```

Comment:
Could probably just be a `ItemStackMetaProvider`.

Reply:
Yeah, I probably should check if there's other specialized `IMetaProvider`s ; hadn't tried that one yet.